File: chat_gemini.py
import os
import google.genai as genai
from google.genai import types
import json
import datetime

def load_schedule():
    try:
        with open("schedule.json", "r") as f:
            return json.load(f)
    except FileNotFoundError:
        return {"events": []}

def save_schedule(schedule):
    with open("schedule.json", "w") as f:
        json.dump(schedule, f, indent=4)

# Define tools for the model
function_declarations = [
    {
        "name": "add_event",
        "description": "Add a new event to the schedule. For recurring events, specify the recurring details.",
        "parameters": {
            "type": "object",
            "properties": {
                "description": {"type": "string", "description": "Description of the event"},
                "date": {"type": "string", "description": "Date of the event (for single events)"},
                "time": {"type": "string", "description": "Time of the event"},
                "recurring": {
                    "type": "object",
                    "description": "Details for recurring events",
                    "properties": {
                        "frequency": {"type": "string", "description": "e.g., weekly"},
                        "days": {"type": "array", "items": {"type": "string"}, "description": "List of days, e.g., ['monday', 'tuesday']"},
                        "count": {"type": "integer", "description": "Number of occurrences"}
                    }
                }
            },
            "required": ["description"]
        }
    },
    {
        "name": "list_events",
        "description": "List all events in the schedule, optionally filtered by date",
        "parameters": {
            "type": "object",
            "properties": {
                "date": {"type": "string", "description": "Optional date filter in YYYY-MM-DD format"}
            }
        }
    },
    {
        "name": "remove_event",
        "description": "Remove an event by index (0-based)",
        "parameters": {
            "type": "object",
            "properties": {"index": {"type": "integer", "description": "Index of the event to remove"}},
            "required": ["index"]
        }
    },
    {
        "name": "get_current_date",
        "description": "Get the current date in YYYY-MM-DD format",
        "parameters": {"type": "object", "properties": {}}
    }
]

tools = [
    types.Tool(function_declarations=function_declarations)
]

# Set up the API key from environment variable
api_key = os.getenv("GEMINI_API_KEY")
if not api_key:
    print("Please set the GEMINI_API_KEY environment variable.")
    exit(1)

client = genai.Client(api_key=api_key)
model = "gemini-2.0-flash"
config = types.GenerateContentConfig(tools=tools)

# Load schedule
schedule = load_schedule()

# Initialize contents with initial prompt
contents = [
    types.Content(
        role="user",
        parts=[
            types.Part(text=f"""You are a personal scheduling assistant. Manage the user's schedule using the available tools. 
Current date: {datetime.date.today().strftime('%Y-%m-%d')}. 
Current schedule: {json.dumps(schedule)}

When the user asks you to decide or choose a time, analyze their current schedule and suggest available time slots that don't conflict with existing events. 
Be proactive in suggesting times based on:
- Avoiding conflicts with existing events
- Common preferences (e.g., morning for exercise, afternoon for meetings)
- Gaps in their schedule
Always provide 2-3 time options when suggesting.

When the user mentions a problem, goal, or improvement area (like "I lack GK", "I need to exercise", "I want to learn coding"), 
be helpful and proactive. Suggest adding relevant events or tasks to their schedule to help them achieve their goal.
For example:
- "I lack GK" → Suggest adding daily/weekly GK reading or quiz sessions
- "I need to exercise" → Suggest adding workout sessions
- "I want to learn X" → Suggest adding study/practice sessions

Always relate their goals back to their schedule and offer to help them make time for improvement.""")
        ],
    ),
]

print("Chat with Gemini Flash 2.0. Type 'exit' to quit.")

while True:
    user_input = input("You: ")
    if user_input.lower() == 'exit':
        break
    contents.append(types.Content(role="user", parts=[types.Part(text=user_input)]))
    
    response = client.models.generate_content(model=model, contents=contents, config=config)
    
    if not response.candidates or not response.candidates[0].content.parts:
        print("Gemini: No response received.")
        continue
    
    for part in response.candidates[0].content.parts:
        if part.text:
            print("Gemini:", part.text)
            contents.append(types.Content(role="model", parts=[types.Part(text=part.text)]))
        if part.function_call:
            call = part.function_call
            result = ""
            if call.name == "add_event":
                desc = call.args.get("description", "")
                date = call.args.get("date", "")
                time = call.args.get("time", "")
                recurring = call.args.get("recurring")
                if not desc:
                    result = "Description is required."
                elif not time:
                    result = "Time is required for all events."
                elif recurring:
                    # Handle recurring
                    frequency = recurring.get("frequency", "weekly")
                    days = recurring.get("days", [])
                    count = recurring.get("count", 4)
                    if not days:
                        result = "Days are required for recurring events."
                    else:
                        today = datetime.date.today()
                        days_map = {"monday": 0, "tuesday": 1, "wednesday": 2, "thursday": 3, "friday": 4, "saturday": 5, "sunday": 6}
                        added = []
                        for day in days:
                            day_num = days_map.get(day.lower())
                            if day_num is not None:
                                days_ahead = (day_num - today.weekday()) % 7
                                if days_ahead == 0:
                                    days_ahead = 7
                                event_date = today + datetime.timedelta(days=days_ahead)
                                for _ in range(count):
                                    schedule["events"].append({"description": desc, "date": event_date.strftime("%Y-%m-%d"), "time": time})
                                    added.append(f"{day} {event_date.strftime('%Y-%m-%d')}")
                                    event_date += datetime.timedelta(days=7)
                        result = f"Added recurring events: {desc} on {', '.join(added)}"
                        save_schedule(schedule)
                else:
                    if not date:
                        result = "Date is required for single events."
                    else:
                        event = {"description": desc, "date": date, "time": time}
                        schedule["events"].append(event)
                        result = f"Added event: {desc} on {date} at {time}"
                        save_schedule(schedule)
            elif call.name == "list_events":
                date_filter = call.args.get("date")
                filtered_events = schedule["events"]
                if date_filter:
                    filtered_events = [e for e in schedule["events"] if e["date"] == date_filter]
                if filtered_events:
                    result = "\n".join([f"{i}. {e['description']} on {e['date']} at {e['time']}" for i, e in enumerate(filtered_events)])
                else:
                    result = f"No events scheduled{' on ' + date_filter if date_filter else ''}."
            elif call.name == "remove_event":
                index = call.args.get("index", -1)
                if 0 <= index < len(schedule["events"]):
                    removed = schedule["events"].pop(index)
                    save_schedule(schedule)
                    result = f"Removed event: {removed['description']}"
                else:
                    result = "Invalid index."
            elif call.name == "get_current_date":
                result = datetime.date.today().strftime("%Y-%m-%d")
            if result:
                # Send the function response
                contents.append(types.Content(
                    role="model",
                    parts=[types.Part(function_response=types.FunctionResponse(name=call.name, response={"result": result}))]
                ))
                # Then, generate follow-up response
                follow_up = client.models.generate_content(model=model, contents=contents, config=config)
                if follow_up.candidates and follow_up.candidates[0].content.parts:
                    for fpart in follow_up.candidates[0].content.parts:
                        if fpart.text:
                            print("Gemini:", fpart.text)
                            contents.append(types.Content(role="model", parts=[types.Part(text=fpart.text)]))
                        # If more calls, but for simplicity, assume one

File: scripts/chat_gemini.ts
import * as fs from 'fs';
import { GoogleGenerativeAI, SchemaType } from '@google/generative-ai';
import * as path from 'path';

interface Event {
  description: string;
  date: string;
  time: string;
}

interface Schedule {
  events: Event[];
}

interface Recurring {
  frequency: string;
  days: string[];
  count: number;
}

interface AddEventArgs {
  description: string;
  date?: string;
  time: string;
  recurring?: Recurring;
}

interface ListEventsArgs {
  date?: string;
}

interface RemoveEventArgs {
  index: number;
}

interface UpdateProfileArgs {
  updates: Record<string, any>;
}

function loadSchedule(): Schedule {
  try {
    const data = fs.readFileSync('schedule.json', 'utf-8');
    return JSON.parse(data);
  } catch {
    return { events: [] };
  }
}

function saveSchedule(schedule: Schedule): void {
  fs.writeFileSync('schedule.json', JSON.stringify(schedule, null, 4));
}

function loadUserProfile(): Record<string, any> {
  try {
    const data = fs.readFileSync('user_profile.json', 'utf-8');
    const profile = JSON.parse(data);
    if ('profile_text' in profile) {
      const newProfile: Record<string, any> = {};
      for (const line of profile.profile_text.split('\n')) {
        if (line.includes(':')) {
          const [key, value] = line.split(':', 2);
          newProfile[key.trim().toLowerCase()] = value.trim();
        }
      }
      return newProfile;
    }
    return profile;
  } catch {
    return {};
  }
}

function saveUserProfile(profile: Record<string, any>): void {
  fs.writeFileSync('user_profile.json', JSON.stringify(profile, null, 4));
}

const functionDeclarations = [
  {
    name: 'add_event',
    description: 'Add a new event to the schedule. For recurring events, specify the recurring details.',
    parameters: {
      type: SchemaType.OBJECT,
      properties: {
        description: { type: SchemaType.STRING },
        date: { type: SchemaType.STRING },
        time: { type: SchemaType.STRING },
        recurring: {
          type: SchemaType.OBJECT,
          properties: {
            frequency: { type: SchemaType.STRING },
            days: { type: SchemaType.ARRAY, items: { type: SchemaType.STRING } },
            count: { type: SchemaType.INTEGER }
          }
        }
      },
      required: ['description']
    }
  },
  {
    name: 'list_events',
    description: 'List all events in the schedule, optionally filtered by date',
    parameters: {
      type: SchemaType.OBJECT,
      properties: {
        date: { type: SchemaType.STRING }
      }
    }
  },
  {
    name: 'remove_event',
    description: 'Remove an event by index (0-based)',
    parameters: {
      type: SchemaType.OBJECT,
      properties: {
        index: { type: SchemaType.INTEGER }
      },
      required: ['index']
    }
  },
  {
    name: 'get_current_date',
    description: 'Get the current date in YYYY-MM-DD format',
    parameters: { type: SchemaType.OBJECT, properties: {} }
  },
  {
    name: 'update_user_profile',
    description: 'Update user profile with any information about the user (name, job, preferences, goals, interests, habits, personality, etc.). Provide updates as key-value pairs.',
    parameters: {
      type: SchemaType.OBJECT,
      properties: {
        updates: {
          type: SchemaType.OBJECT
        }
      },
      required: ['updates']
    }
  },
  {
    name: 'get_user_profile',
    description: 'Get the current user profile information',
    parameters: { type: SchemaType.OBJECT, properties: {} }
  }
];

class ChatSession {
  schedule: Schedule;
  userProfile: Record<string, any>;
  contents: any[];

  constructor() {
    this.schedule = loadSchedule();
    this.userProfile = loadUserProfile();
    this.contents = this.initializeContents();
  }

  private initializeContents() {
    const today = new Date().toISOString().split('T')[0];
    return [
      {
        role: 'user',
        parts: [
          {
            text: `You are a personal scheduling assistant. Manage the user's schedule using the available tools. 
Current date: ${today}. 
Current schedule: ${JSON.stringify(this.schedule)}
User profile: ${JSON.stringify(this.userProfile)}

IMPORTANT: Never mention tool calls, function calls, or show tool outputs in your responses. Just respond naturally based on the results.

When the user asks you to decide or choose a time, analyze their current schedule and suggest available time slots that don't conflict with existing events. 
Be proactive in suggesting times based on:
- Avoiding conflicts with existing events
- User preferences from their profile (morning_start, evening_end, preferred times)
- Common preferences (e.g., morning for exercise, afternoon for meetings)
- Gaps in their schedule
Always provide 2-3 time options when suggesting.

When the user mentions a problem, goal, or improvement area (like "I lack GK", "I need to exercise", "I want to learn coding"), 
be helpful and proactive. Suggest adding relevant events or tasks to their schedule to help them achieve their goal.
For example:
- "I lack GK" → Suggest adding daily/weekly GK reading or quiz sessions
- "I need to exercise" → Suggest adding workout sessions
- "I want to learn X" → Suggest adding study/practice sessions

When the user shares personal information (name, job, preferences, goals, interests, habits, personality traits, etc.), 
silently update their profile using update_user_profile without mentioning it in your response.
Use the user profile to personalize your responses and suggestions based on what you know about them.

Always relate their goals back to their schedule and offer to help them make time for improvement.`
          }
        ]
      }
    ];
  }

  addUserMessage(userInput: string): void {
    this.contents.push({ role: 'user', parts: [{ text: userInput }] });
  }

  addModelResponse(text: string): void {
    this.contents.push({ role: 'model', parts: [{ text }] });
  }

  addFunctionResponse(callName: string, result: any): void {
    this.contents.push({
      role: 'model',
      parts: [{ functionResponse: { name: callName, response: { result } } }]
    });
  }

  handleFunctionCall(call: any): string {
    let result = '';
    if (call.name === 'add_event') {
      const args: AddEventArgs = call.args;
      const desc = args.description;
      const date = args.date;
      const time = args.time;
      const recurring = args.recurring;
      if (!desc) {
        result = 'Description is required.';
      } else if (!time) {
        result = 'Time is required for all events.';
      } else if (recurring) {
        const frequency = recurring.frequency || 'weekly';
        const days = recurring.days;
        const count = recurring.count || 4;
        if (!days || days.length === 0) {
          result = 'Days are required for recurring events.';
        } else {
          const today = new Date();
          const daysMap: Record<string, number> = { monday: 1, tuesday: 2, wednesday: 3, thursday: 4, friday: 5, saturday: 6, sunday: 0 };
          const added: string[] = [];
          for (const day of days) {
            const dayNum = daysMap[day.toLowerCase()];
            if (dayNum !== undefined) {
              let daysAhead = (dayNum - today.getDay() + 7) % 7;
              if (daysAhead === 0) daysAhead = 7;
              const eventDate = new Date(today);
              eventDate.setDate(today.getDate() + daysAhead);
              for (let i = 0; i < count; i++) {
                this.schedule.events.push({ description: desc, date: eventDate.toISOString().split('T')[0], time });
                added.push(`${day} ${eventDate.toISOString().split('T')[0]}`);
                eventDate.setDate(eventDate.getDate() + 7);
              }
            }
          }
          result = `Added recurring events: ${desc} on ${added.join(', ')}`;
          saveSchedule(this.schedule);
        }
      } else {
        if (!date) {
          result = 'Date is required for single events.';
        } else {
          this.schedule.events.push({ description: desc, date, time });
          result = `Added event: ${desc} on ${date} at ${time}`;
          saveSchedule(this.schedule);
        }
      }
    } else if (call.name === 'list_events') {
      const args: ListEventsArgs = call.args;
      const dateFilter = args.date;
      let filteredEvents = this.schedule.events;
      if (dateFilter) {
        filteredEvents = this.schedule.events.filter(e => e.date === dateFilter);
      }
      if (filteredEvents.length > 0) {
        result = filteredEvents.map((e, i) => `${i}. ${e.description} on ${e.date} at ${e.time}`).join('\n');
      } else {
        result = `No events scheduled${dateFilter ? ' on ' + dateFilter : ''}.`;
      }
    } else if (call.name === 'remove_event') {
      const args: RemoveEventArgs = call.args;
      const index = args.index;
      if (index >= 0 && index < this.schedule.events.length) {
        const removed = this.schedule.events.splice(index, 1)[0];
        saveSchedule(this.schedule);
        result = `Removed event: ${removed.description}`;
      } else {
        result = 'Invalid index.';
      }
    } else if (call.name === 'get_current_date') {
      result = new Date().toISOString().split('T')[0];
    } else if (call.name === 'update_user_profile') {
      const args: UpdateProfileArgs = call.args;
      const updates = args.updates;
      Object.assign(this.userProfile, updates);
      saveUserProfile(this.userProfile);
      result = 'Profile updated';
    } else if (call.name === 'get_user_profile') {
      result = this.userProfile ? JSON.stringify(this.userProfile) : 'No profile information available.';
    }

    if (result) {
      this.addFunctionResponse(call.name, result);
    }
    return result;
  }
}

export { ChatSession, loadSchedule, saveSchedule, loadUserProfile, saveUserProfile, functionDeclarations };
